ktorm-core/src/main/kotlin/me/liuwj/ktorm/database/SqlDialect.kt

```diff
@@ -45,7 +48,52 @@ interface SqlDialect {
      * @param indentSize the indent size.
      * @return a [SqlFormatter] object, generally typed of subclasses to support dialect-specific sql expressions.
      */
-    fun createSqlFormatter(database: Database, beautifySql: Boolean, indentSize: Int): SqlFormatter
+    fun createSqlFormatter(database: Database, beautifySql: Boolean, indentSize: Int): SqlFormatter {
+        return SqlFormatter(database, beautifySql, indentSize)
+    }
+
+    /**
+     * Format the provided [SqlExpression] to a SQL string with its execution arguments, then create
+     * a [PreparedStatement] from the global database using the SQL string and execute the specific
+     * callback function with it. After the callback function completes, the statement will be
+     * closed automatically.
+     *
+     * The default implementation simply calls the inline function [SqlExpression.prepareStatement]. Specific
+     * dialects may override this if the default technique is unimplemented.
+     *
+     * @param expression The expression to prepare as a statement
+     * @param autoGeneratedKeys a flag indicating whether auto-generated keys should be made available for retrieval.
+     * @param func the executed callback function.
+     * @return the result of the callback function.
+     */
+    fun <T> prepareStatement(
```

Comment:
Would it be better to modify this as follows? Then we just need to add one function to `SqlDialect`: 

```kotlin
fun <T> executeAndGetGeneratedKeys(
    database: Database,
    sql: String,
    args: List<ArgumentExpression<*>>,
    resultSetHandler: (ResultSet) -> T
): T {
    database.useConnection { conn ->
        conn.prepareStatement(sql, Statement.NO_GENERATED_KEYS).use { statement ->
            statement.setArguments(args)
            statement.executeUpdate()
            return resultSetHandler(statement.generatedKeys)
        }
    }
}
```

Reply:
Yes, I like that idea, with some tweaking:

Assuming that:

*  only one auto-generated key is possible (I am not aware of an SQL implementations that allow more) 
*  an autogenerated key must be a number (again I'm not aware of any other implementations):
* Only an insert operation can autogenerate a key (correct me if I'm wrong)
Then this would be how I would do it. This is the default implementation in `SqlDialect`

````
    fun insertAndGenerateKey(
        conn: Connection,
        sql: String,
        args: List<ArgumentExpression<*>>
    ): Number {
        conn.prepareStatement(sql, Statement.RETURN_GENERATED_KEYS).use { statement ->
            statement.use {
                statement.setArguments(args)
                statement.executeUpdate()
                return statement.generatedKeys.use { rs ->
                    if (rs.next()) {
                        return rs.getObject(1) as? Number ?: error("Generated key is not a number")
                    } else {
                        error("No key generated by insert.")
                    }
                }
            }
        }
    }
````
And the `SqLiteDialect` version (which would work on *all* SQLite implementations so no need for `SimpleSqLiteDialect`) would be:
````
    override fun insertAndGenerateKey(
        conn: Connection,
        sql: String,
        args: List<ArgumentExpression<*>>
    ): Number {
        conn.prepareStatement(sql).use { statement ->
            statement.use {
                statement.setArguments(args)
                statement.executeUpdate()
            }
        }
        return conn.prepareStatement("select last_insert_rowid()").use { statement ->
            val rs = statement.executeQuery()
            if (rs.next()) {
                return rs.getObject(1) as? Number ?: error("Generated key is not a number")
            } else {
                error("No key generated by insert.")
            }
        }
    }
````
And the extension function `BaseTable.insertAndGenerateKey()` becomes:
````
fun <T : BaseTable<*>> T.insertAndGenerateKey(block: AssignmentsBuilder.(T) -> Unit): Any {
    val assignments = ArrayList<ColumnAssignmentExpression<*>>()
    AssignmentsBuilder(assignments).block(this)

    val expression = AliasRemover.visit(InsertExpression(asExpression(), assignments))
    return with(Database.global) {
        val (sql, args) = formatExpression(expression)
        useConnection { connection ->
            dialect.insertAndGenerateKey(connection, sql, args)
        }
    }
}
````
And a new extension function
````
fun PreparedStatement.setArguments(args: List<ArgumentExpression<*>>) {
    args.forEachIndexed { i, expr ->
        @Suppress("UNCHECKED_CAST")
        val sqlType = expr.sqlType as SqlType<Any>
        sqlType.setParameter(this, i + 1, expr.value)
    }
}
````
I've implemented and tested that.

Also, my PR made the `dialect` property of `Database` non-null, since an implementation is always needed. Are you ok with that?

I haven't yet put back the debug logging, largely because the optional `logger` property bugs me - code like 
````
        val logger = Database.global.logger
        if (logger != null && logger.isDebugEnabled()) {
            logger.debug("Effects: $effects")
        }
````
is ugly. It would be nicer if `logger` was non optional, with a default stubbed implementation, and if enabling debug was internal to the logger, so the `isDebugEnabled()` call was not required, i.e. you could just write `Database.global.logger.debug("Effects: $effects")`.

But I'll deal with that in a separate issue.

Reply:
Oh, and an additional function might be useful for batched inserts.

Reply:
I don't think we should make these assumptions. There are few exceptions, but they do exists. So it'd be better to process the result set outside `SqlDialect`. 

In my opinion, the default implementation of this function in `SqlDialect` should be:

```kotlin
fun <T> executeAndGetGeneratedKeys(
    database: Database,
    sql: String,
    args: List<ArgumentExpression<*>>,
    resultSetHandler: (ResultSet) -> T
): T {
    database.useConnection { conn ->
        conn.prepareStatement(sql, Statement.NO_GENERATED_KEYS).use { statement ->
            statement.setArguments(args)
            statement.executeUpdate()
            return resultSetHandler(statement.generatedKeys)
        }
    }
}
``` 

The SQLite version: 

```kotlin
override fun <T> executeAndGetGeneratedKeys(
    database: Database,
    sql: String,
    args: List<ArgumentExpression<*>>,
    resultSetHandler: (ResultSet) -> T
): T {
    database.useConnection { conn ->
        conn.prepareStatement(sql).use { statement ->
            statement.setArguments(args)
            statement.executeUpdate()

            conn.prepareStatement("select last_insert_rowid()").use { query ->
                return resultSetHandler(query.executeQuery())
            }
        }
    }
}
```

Then the extension function `BaseTable.insertAndGenerateKey` becomes: 

```kotlin
fun <T : BaseTable<*>> T.insertAndGenerateKey(block: AssignmentsBuilder.(T) -> Unit): Any {
    val assignments = ArrayList<ColumnAssignmentExpression<*>>()
    AssignmentsBuilder(assignments).block(this)

    val expression = AliasRemover.visit(InsertExpression(asExpression(), assignments))

    with(Database.global) {
        val (sql, args) = formatExpression(expression)

        return dialect!!.executeAndGetGeneratedKeys(this, sql, args) { rs ->
            if (rs.next()) {
                val sqlType = primaryKey?.sqlType ?: error("Table $tableName must have a primary key.")
                sqlType.getResult(rs, 1) ?: error("Generated key is null.")
            } else {
                error("No generated key returns by database.")
            }
        }
    }
}
```

Reply:
Well, assumptions are fine as long as they're clearly stated. But your proposal works though I still don't see the need to pass the database to the dialect function - it's better to wrap the whole process inside a single `useConnection` call so that the same connection is used where a separate query is required to fetch they key. I'll update the PR accordingly for your review.